Added NFT minting functionality to Quiz contract, quiz now displays s…

ZigaMr opened the pull request:
…tatic NFT image after winning. Expect to add IPFS storage and auto-generated NFTs.

Implements #2

matevz commented on backend/contracts/Quiz.sol:
```diff
@@ -3,16 +3,19 @@ pragma solidity ^0.8.0;
 
-contract Quiz {
+
+contract Quiz is ERC721 {
```
Let's decouple Quiz and ERC721. Add `setNft()` (and  `getNft()`) with parameters like the NFT name, symbol and image URI which will instantiate and reference a new ERC721. Then in the claimReward(), the function checks, whether the NFT was initialized and performs the NFT minting.

ZigaMr replied:
Sure, added separate contract for the ERC721, now stores URIs at constructor level.

matevz commented on backend/contracts/Quiz.sol:
```diff
@@ -3,16 +3,19 @@ pragma solidity ^0.8.0;
 import "@oasisprotocol/sapphire-contracts/contracts/Sapphire.sol";
 import {EIP155Signer} from "@oasisprotocol/sapphire-contracts/contracts/EIP155Signer.sol";
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+// import "@openzeppelin/contracts/access/Ownable.sol";
```
leftover?

ZigaMr replied:
Yes, removed it now.

matevz commented on backend/contracts/Quiz.sol:
```diff
@@ -3,16 +3,19 @@ pragma solidity ^0.8.0;
     string constant errPayoutFailed = "Payout failed";
 
-    uint256 public constant COUPON_VALID =  type(uint256).max-1;
+    uint256 public constant COUPON_VALID = type(uint256).max-1;
```
I see you're using a formatter. Can you add it and the linter to the package.json?

See for example https://github.com/oasisprotocol/sapphire-paratime/blob/main/contracts/package.json#L13-L20.

And then require the linter to pass in the github actions: https://github.com/oasisprotocol/sapphire-paratime/blob/main/.github/workflows/ci-lint.yaml#L81

This can go into a separate PR.

ZigaMr replied:
Sure, will add to a separate PR.

ZigaMr committed:
Added separate NFT contract for minting ERC721 tokens, removed linter, kept frontend changes

Cosmetic changes

Add SVG base64 image to NFT contract state, tests for OasisReward contract, update hardhat.config.ts and fix CI/CD workflows.

Add automatically generated base64 image to NFT contract state, tests for NFT-OasisReward contract, update hardhat.config to include NFT deploy and setup

Add separate tests for Rewards contract, increase gasless tx gasLimit parameter, add docBlock comments

Add CI/CD update, cherry-pick whitelist commit from backup branch.

Cosmetic changes

Fix tests for CI/CD

Update ci-test.yaml

Add SVG base64 image to NFT contract state, tests for OasisReward contract, update hardhat.config.ts and fix CI/CD workflows.

Add automatically generated base64 image to NFT contract state, tests for NFT-OasisReward contract, update hardhat.config to include NFT deploy and setup

Add separate tests for Rewards contract, increase gasless tx gasLimit parameter, add docBlock comments

Add CI/CD update, cherry-pick whitelist commit from backup branch.

Cosmetic changes

Fix tests for CI/CD

Update ci-test.yaml

Add Chain ID check to tests, add reward flags for NFT/ROSE

Fix frontend/package.json

Remove pnpm-lock.yaml

Remove vscode settings, add natspec

matevz commented on backend/contracts/Quiz.sol:
```diff
@@ -273,12 +321,29 @@ contract Quiz {
+    // Function to deploy a new ERC721 contract and set its address
+    function setNft(string memory name, string memory symbol, string[] memory ipfsURIs) public {
+        QuizERC721 newNft = new QuizERC721(name, symbol, ipfsURIs);
+        nftContractAddress = address(newNft);
```
```suggestion
        nftContractAddress = new QuizERC721(name, symbol, ipfsURIs);
```

matevz commented on backend/contracts/Quiz.sol:
```diff
@@ -3,8 +3,56 @@ pragma solidity ^0.8.0;
+    function mint(address to, uint256 tokenId) public {
+        require(bytes(_tokenURIs[tokenId]).length > 0, "Token URI not set");
+        _safeMint(to, tokenId);
+    }
```
```suggestion
    /**
      * @notice Mint new NFT reward.
      * @param to ...
      * @param ipfsUri ...
      */
    function mint(address to, string calldata ipfsUri) public {
        require(ipfsUri.length > 0, "Token URI not set");
        uint256 tokenId = totalSupply();
        _tokenURIs[tokenId] = ipfsUri;
        _safeMint(to, tokenId);
    }
```

matevz commented on backend/contracts/Quiz.sol:
```diff
@@ -3,8 +3,56 @@ pragma solidity ^0.8.0;
 import {EIP155Signer} from "@oasisprotocol/sapphire-contracts/contracts/EIP155Signer.sol";
+import "@openzeppelin/contracts/token/ERC721/extensions/ERC721Enumerable.sol";
+
+contract QuizERC721 is ERC721Enumerable {
```
```suggestion
contract OasisReward is ERC721Enumerable {
```

matevz replied:
Also move to a separate file.

matevz commented on backend/contracts/Quiz.sol:
```diff
@@ -3,8 +3,56 @@ pragma solidity ^0.8.0;
+    // Mapping from user address to list of owned token IDs
+    mapping(address => uint256[]) private _ownedTokens;
+
+    constructor(string memory name, string memory symbol, string[] memory ipfsURIs) ERC721(name, symbol) {
```
```suggestion
    constructor(string memory name, string memory symbol) ERC721(name, symbol) {
```

matevz reviewed (commented):
TODO:
- hardhat task for setting an NFT and initializing a new NFT contract, if needed
- tests

matevz commented on backend/contracts/OasisReward.sol:
```diff
@@ -0,0 +1,84 @@
+     * @param base64EncodedSVG base64 encoded SVG image
+     */
+    // Function to mint new tokens
+    function mint(address to, string calldata base64EncodedSVG) public onlyOwner {
```
Is `onlyOwner` the correct modifier here? I think the caller (quiz) should be checked, if it's whitelisted, because you can have multiple quizzes using the same NFT contract. Sth like:

```suggestion
    mapping(address => bool) _allowMint;
    
    modifier onlyAllowMint {
      require(_allowMint[msg.sender], "not allowed to mint");
      _;
    }
    
    function addAllowMint(address) onlyOwner {
      ...
    }
    
    function removeAllowMint(address) onlyOwner {
      ...
    }
    
    function mint(address to, string calldata base64EncodedSVG) public onlyAllowMint {
```

ZigaMr replied:
You're right, the tx.origin check does not take into account user mints and gasless tx, will change to whitelisted addresses.

matevz reviewed (commented):
Instead of merging the main branch https://github.com/oasisprotocol/demo-quiz/pull/3/files/93d55a1d09da32d5d81ce018a9a7dfd66976a78d..2a44bed3c185580d9847746aabe7ab64d3dfeb44 simply rebase your PR to the main branch.

matevz commented on .vscode/settings.json:
```diff
@@ -0,0 +1,2 @@
+{
```
Remove from git.

matevz commented on backend/contracts/OasisReward.sol:
```diff
@@ -0,0 +1,116 @@
+import "@openzeppelin/contracts/utils/Base64.sol";
+
+contract OasisReward is ERC721Enumerable {
+    // Mapping from token ID to IPFS URI
```
Please use the [natspec](https://docs.soliditylang.org/en/latest/natspec-format.html) format where possible. Same for other variables.

```suggestion
    /// Mapping from token ID to IPFS URI.
```

ZigaMr replied:
Added natspec format to OasisReward.sol and deleted vscode settings file.

matevz reviewed (commented):
Some nits.
